### PR TITLE
refactor(amplify-codegen-appsync-model-plugin): auth model annotation

### PR DESCRIPTION
**Pull request by nikhname**
*Issue #, if available:*
#4509 
*Description of changes:*
Add auth model annotations according to https://github.com/aws-amplify/amplify-android/pull/559

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap [hunk 1]
```diff
@@ -1,5 +1,1771 @@
...
+import java.util.UUID;
+import java.util.Objects;
+
+import androidx.core.util.ObjectsCompat;
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.annotations.Index;
+import com.amplifyframework.core.model.annotations.ModelConfig;
+import com.amplifyframework.core.model.annotations.ModelField;
+import com.amplifyframework.core.model.query.predicate.QueryField;
```
Comment (richardmcclellan) [inline packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap:213]: We also need to make sure the following import statements are added, if used:

import com.amplifyframework.core.model.AuthStrategy;
import com.amplifyframework.core.model.ModelOperation;
import com.amplifyframework.core.model.annotations.AuthRule;

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 2]
```diff
@@ -753,6 +761,55 @@ export class AppSyncModelJavaVisitor<
...
+        const authRule = [];
+        if (rule.allow === AuthStrategy.groups) {
+          authRule.push('allow = AuthStrategy.GROUPS');
+          if (rule.groupClaim) {
+            authRule.push(`groupClaim = "${rule.groupClaim}"`);
+          }
+          if (rule.groups) {
+            authRule.push(`groups = { ${rule.groups?.map(group => `"${group}"`).join(', ')} }`);
+          } else {
+            authRule.push(`groupsField = "${rule.groupField}"`);
```
Comment (richardmcclellan) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:777]: I don't think we should assume that if `groups` is not present, `groupsField` exists.  Instead, we should just do something like: 
```
if (rule.groups) {
    authRule.push(`groups = { ${rule.groups?.map(group => `"${group}"`).join(', ')} }`);
}
if (rule.groupsField) {
    authRule.push(`groupsField = "${rule.groupField}"`);
}
```

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 3]
```diff
@@ -753,6 +761,55 @@ export class AppSyncModelJavaVisitor<
...
+              authRule.push('allow = AuthStrategy.OWNER');
+              break;
+            case AuthStrategy.private:
+              authRule.push('allow = AuthStrategy.PRIVATE');
+              break;
+            case AuthStrategy.public:
+              authRule.push('allow = AuthStrategy.PUBLIC');
+              break;
+            default:
+              printWarning(`Model has auth with authStrategy ${rule.allow} of which is not yet supported in DataStore.`);
```
Comment (richardmcclellan) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:791]:
```suggestion
              printWarning(`Model has auth with authStrategy ${rule.allow} of which is not yet supported in Amplify.`);
```
The @auth directive isn't really specific to DataStore, it affects both the API and DataStore categories, so I would suggest saying this isn't supported in Amplify.

File: .circleci/config.yml [hunk 4]
```diff
@@ -1024,12 +1024,12 @@ workflows:
           filters: *ref_2
           requires:
             - publish_to_local_registry
-            - layer-amplify_e2e_tests
+            - hostingPROD-amplify_e2e_tests
```
Comment (richardmcclellan) [inline .circleci/config.yml:1027]: Is this change related to supporting @auth directive?  Is it intentional?

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 5]
```diff
@@ -737,13 +739,19 @@ export class AppSyncModelJavaVisitor<
     const annotations: string[] = model.directives.map(directive => {
       switch (directive.name) {
         case 'model':
-          return `ModelConfig(pluralName = "${this.pluralizeModelName(model)}")`;
-          break;
+          const modelArgs: string[] = [];
+          const authDirectives: AuthDirective[] = (model.directives.filter(d => d.name === 'auth') as any) as AuthDirective[];
```
Comment (akshbhu) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:743]: Can we use 
```
(model.directives.filter(d => d.name === 'auth') as CodeGenDirective)
```
instead of any

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 6]
```diff
@@ -761,7 +818,13 @@ export class AppSyncModelJavaVisitor<
   }
 
   protected generateModelFieldAnnotation(field: CodeGenField): string {
-    const annotationArgs: string[] = [`targetType="${field.type}"`, !field.isNullable ? 'isRequired = true' : ''].filter(arg => arg);
+    const authDirectives: AuthDirective[] = (field.directives.filter(d => d.name === 'auth') as any) as AuthDirective[];
```
Comment (akshbhu) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:821]: same as above comment line 743

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 7]
```diff
@@ -761,7 +818,13 @@ export class AppSyncModelJavaVisitor<
...
+    const authRules = this.generateAuthRules(authDirectives);
+    const annotationArgs: string[] = [
+      `targetType="${field.type}"`,
+      !field.isNullable ? 'isRequired = true' : '',
+      authRules.length ? `authRules = ${authRules}` : '',
+    ].filter(arg => arg);
```
Comment (akshbhu) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:827]:
``` javascript
.filter(arg => arg);
```
this condition will return same array right?

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 5 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:743]: Sure, d is already of type CodeGenDirective so we can remove the `as any`.

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 7 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:827]: Yup will remove this filter

Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:827]: This code was already present, I believe it is there for formatting purposes (checking for string type)

File: .circleci/config.yml (hunk 4 above)
Comment (nikhname) [reply .circleci/config.yml:1027]: This change is unintentional, removing

Comment (nikhname) [reply .circleci/config.yml:1027]: This change is coming from a commit hook script, master will need to be updated first

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 2 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts]: Made this change

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 3 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts]: Changed to "not yet supported"

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 6 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts]: Removed `as any`

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 8]
```diff
@@ -753,6 +763,67 @@ export class AppSyncModelJavaVisitor<
...
+            authRule.push('allow = AuthStrategy.PRIVATE');
+            if (rule.ownerField) {
+              authRule.push(`ownerField = "${rule.ownerField}"`);
+            }
+            if (rule.identityClaim) {
+              authRule.push(`identityClaim = "${rule.identityClaim}"`);
+            }
+            break;
+          case AuthStrategy.public:
+            authRule.push('allow = AuthStrategy.PUBLIC');
```
Comment (richardmcclellan) [inline packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:791]: `ownerField` and `identityClaim` only apply to `AuthStrategy.OWNER`, so you can remove them from the `AuthStrategy.PRIVATE` and `AuthStrategy.PUBLIC` cases.

File: packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap [hunk 9]
```diff
@@ -1374,9 +1395,12 @@ public final class privateType implements Model {
 "
 `;
 
-exports[`AppSyncModelVisitor Model with Auth should generate models with public authorization 1`] = `
+exports[`AppSyncModelVisitor Model with Auth should generate class with public authorization 1`] = `
 "package com.amplifyframework.datastore.generated.model;
 
+import com.amplifyframework.core.model.AuthStrategy;
```
Comment (richardmcclellan) [inline packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap:1401]: Can you make sure imports are sorted alphabetically?  These should be in the section below with the other `com.amplifyframework.core.model.` imports.

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 8 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:795]: Removed ownerField and identityClaim

File: packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap (hunk 9 above)
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap]: Moved the imports to be with the other com.amplifyframework.core.model imports

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 10]
```diff
@@ -263,7 +267,16 @@ export class AppSyncModelJavaVisitor<
   }
 
   protected generatePackageHeader(isModel: boolean = true): string {
-    const baseImports = isModel ? MODEL_CLASS_IMPORT_PACKAGES : NON_MODEL_CLASS_IMPORT_PACKAGES;
+    let baseImports;
+    if (isModel) {
+      if (this.usingAuth) {
+        baseImports = MODEL_AUTH_CLASS_IMPORT_PACKAGES;
```
[the remark this answers is not in the document]
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:273]: Modified the constant generation to use a separate list for the Auth imports, which are merged into the Model imports.

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts [hunk 11]
```diff
@@ -753,6 +773,55 @@ export class AppSyncModelJavaVisitor<
...
+  protected generateAuthRules(authDirectives: AuthDirective[]): string {
+    const rules: string[] = [];
+    authDirectives.forEach(directive => {
+      directive.arguments?.rules.forEach(rule => {
+        const authRule = [];
+        switch (rule.allow) {
+          case AuthStrategy.owner:
+            authRule.push('allow = AuthStrategy.OWNER');
+            if (rule.ownerField) {
+              authRule.push(`ownerField = "${rule.ownerField}"`);
```
[the remark this answers is not in the document]
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:785]: What do you mean be keeping this information in the CLI?

File: packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (nikhname) [reply packages/amplify-codegen-appsync-model-plugin/src/visitors/appsync-java-visitor.ts:273]: https://github.com/aws-amplify/amplify-cli/blob/a0a53f05ac98d02b0a3ecf86b71f2f447169b1ee/packages/amplify-codegen-appsync-model-plugin/src/configs/java-config.ts#L29
